File: src/processors/detection/base/detection_pass.py
from abc import abstractmethod

from src.processors.detection.base.common_pass import FilePassAggregates
from src.processors.detection.base.detection import FieldDetection
from src.processors.layout.field.base import Field
from src.utils.logger import Logger
from src.utils.stats import StatsByLabel


class FieldTypeDetectionPass(FilePassAggregates):
    """FieldTypeDetectionPass implements detection pass for specific field types, managing the detection-related aggregates.".

    It is responsible for executing detection logic on the image from the field information.
    It does not determine the actual field values, that is left to the interpretation pass
    which can make use of aggregate data collected during the detection pass.
    """

    def __init__(self, tuning_config, field_detection_type) -> None:
        self.field_detection_type = field_detection_type
        super().__init__(tuning_config)

    @abstractmethod
    def get_field_detection(
        self, field: Field, gray_image, colored_image
    ) -> FieldDetection:
        # Not implemented
        raise NotImplementedError

    def update_aggregates_on_processed_field_detection(
        self, field: Field, field_detection: FieldDetection
    ) -> None:
        self.update_field_level_aggregates_on_processed_field_detection(
            field, field_detection
        )

        field_level_aggregates = self.get_field_level_aggregates()

        self.update_file_level_aggregates_on_processed_field_detection(
            field, field_detection, field_level_aggregates
        )

        self.update_directory_level_aggregates_on_processed_field_detection(
            field, field_detection, field_level_aggregates
        )

    def update_field_level_aggregates_on_processed_field_detection(
        self, field, _field_detection
    ) -> None:
        super().update_field_level_aggregates_on_processed_field(field)

    def update_file_level_aggregates_on_processed_field_detection(
        self, field, _field_detection, field_level_aggregates
    ) -> None:
        super().update_file_level_aggregates_on_processed_field(
            field, field_level_aggregates
        )
        # detection confidence metrics?

    def update_directory_level_aggregates_on_processed_field_detection(
        self, field, _field_detection, field_level_aggregates
    ) -> None:
        super().update_directory_level_aggregates_on_processed_field(
            field, field_level_aggregates
        )
        # (if needed) update_directory_level_aggregates_on_processed_field


class TemplateDetectionPass(FilePassAggregates):
    def initialize_directory_level_aggregates(
        self, initial_directory_path, all_field_detection_types
    ) -> None:
        super().initialize_directory_level_aggregates(initial_directory_path)
        self.insert_directory_level_aggregates(
            {
                "files_by_label_count": StatsByLabel("processed", "multi_marked"),
                "field_detection_type_wise_aggregates": {
                    key: {"fields_count": StatsByLabel("processed")}
                    for key in all_field_detection_types
                },
            }
        )

    def initialize_file_level_aggregates(
        self, file_path, all_field_detection_types
    ) -> None:
        super().initialize_file_level_aggregates(file_path)
        self.insert_file_level_aggregates(
            {
                "field_detection_type_wise_aggregates": {
                    key: {"fields_count": StatsByLabel("processed")}
                    for key in all_field_detection_types
                },
            }
        )

    def update_aggregates_on_processed_field_detection(
        self, field: Field, field_detection: FieldDetection
    ) -> None:
        self.update_field_level_aggregates_on_processed_field_detection(
            field, field_detection
        )

        # TODO: can we move this getter to parent class?
        field_level_aggregates = self.get_field_level_aggregates()

        self.update_file_level_aggregates_on_processed_field_detection(
            field, field_detection, field_level_aggregates
        )
        self.update_directory_level_aggregates_on_processed_field_detection(
            field, field_detection, field_level_aggregates
        )

    def update_field_level_aggregates_on_processed_field_detection(
        self, field, _field_detection
    ) -> None:
        super().update_field_level_aggregates_on_processed_field(field)

    def update_file_level_aggregates_on_processed_field_detection(
        self, field, _field_detection, field_level_aggregates
    ) -> None:
        super().update_file_level_aggregates_on_processed_field(
            field, field_level_aggregates
        )
        # detection confidence metrics?

    def update_directory_level_aggregates_on_processed_field_detection(
        self, field: Field, _field_detection, field_level_aggregates
    ) -> None:
        super().update_directory_level_aggregates_on_processed_field(
            field, field_level_aggregates
        )
        field_detection_type = field.field_detection_type

        field_detection_type_wise_aggregates = self.directory_level_aggregates[
            "field_detection_type_wise_aggregates"
        ][field_detection_type]
        # Update the processed field count for that runner
        field_detection_type_wise_aggregates["fields_count"].push("processed")

    # TODO: check if passing runners is really needed or not here -
    def update_aggregates_on_processed_file(
        self, file_path, field_detection_type_file_runners
    ) -> None:
        super().update_aggregates_on_processed_file(file_path)

        field_detection_type_wise_aggregates = self.file_level_aggregates[
            "field_detection_type_wise_aggregates"
        ]
        for (
            field_detection_type_file_runner
        ) in field_detection_type_file_runners.values():
            field_detection_type_wise_aggregates[
                field_detection_type_file_runner.field_detection_type
            ] = field_detection_type_file_runner.get_file_level_detection_aggregates()

        # Populate typed field results from repository
        # Get repository from any field type runner (they all share the same repository)
        any_runner = next(iter(field_detection_type_file_runners.values()), None)
        if any_runner and hasattr(any_runner, "repository") and any_runner.repository:
            try:
                file_results = any_runner.repository.get_file_results(file_path)

                # Map all field types by field_label for interpretation access
                bubble_fields_by_label = {
                    result.field_label: result
                    for result in file_results.bubble_fields.values()
                }
                ocr_fields_by_label = {
                    result.field_label: result
                    for result in file_results.ocr_fields.values()
                }
                barcode_fields_by_label = {
                    result.field_label: result
                    for result in file_results.barcode_fields.values()
                }

                self.file_level_aggregates["bubble_fields"] = bubble_fields_by_label
                self.file_level_aggregates["ocr_fields"] = ocr_fields_by_label
                self.file_level_aggregates["barcode_fields"] = barcode_fields_by_label
            except KeyError as e:
                # File not yet finalized in repository - this should not happen
                # but if it does, log and re-raise to surface the issue
                logger = Logger(__name__)
                logger.error(
                    f"File {file_path} not found in repository after finalize_file(). "
                    f"This indicates a bug in the repository lifecycle management. Error: {e}"
                )
                # Re-raise to surface the issue during development
                raise
            except Exception as e:
                # Other unexpected errors
                logger = Logger(__name__)
                logger.error(
                    f"Unexpected error populating field results from repository for {file_path}: {e}"
                )
                # Initialize empty dicts to prevent KeyError in interpretation
                self.file_level_aggregates.setdefault("bubble_fields", {})
                self.file_level_aggregates.setdefault("ocr_fields", {})
                self.file_level_aggregates.setdefault("barcode_fields", {})

File: omrchecker-js/packages/core/src/processors/detection/base/detectionPass.ts
/**
 * Detection pass base classes.
 *
 * TypeScript port of src/processors/detection/base/detection_pass.py
 * Defines abstract base classes for field type detection passes and template-level detection passes.
 */

import type { Field } from '../../layout/field/base';
import { StatsByLabel } from '../../../utils/stats';
import {
  FilePassAggregates,
  type FieldLevelAggregates,
  type TuningConfig,
} from './commonPass';
import type { FieldDetection } from './detection';
import { DetectionRepository } from '../../repositories/DetectionRepository';
import {
  BubbleFieldDetectionResult,
  OCRFieldDetectionResult,
  BarcodeFieldDetectionResult,
} from '../models';
import { Logger } from '../../../utils/logger';

/**
 * Abstract base class for field type detection passes.
 *
 * FieldTypeDetectionPass implements detection pass for specific field types,
 * managing the detection-related aggregates.
 *
 * It is responsible for executing detection logic on the image from the field information.
 * It does not determine the actual field values, that is left to the interpretation pass
 * which can make use of aggregate data collected during the detection pass.
 */
export abstract class FieldTypeDetectionPass extends FilePassAggregates {
  public fieldDetectionType: string;

  constructor(tuningConfig: TuningConfig, fieldDetectionType: string) {
    // Call super() first (proper initialization order)
    super(tuningConfig);
    // Set subclass-specific properties after super()
    this.fieldDetectionType = fieldDetectionType;
  }

  /**
   * Abstract method to get field detection.
   * Must be implemented by subclasses.
   *
   * @param field - Field to detect
   * @param grayImage - Grayscale image
   * @param coloredImage - Colored image (optional)
   * @returns FieldDetection result
   */
  abstract getFieldDetection(
    field: Field,
    grayImage: cv.Mat,
    coloredImage?: cv.Mat
  ): FieldDetection;

  /**
   * Run field-level detection with automatic initialization of aggregates.
   *
   * @param field - Field to detect
   * @param grayImage - Grayscale image
   * @param coloredImage - Colored image (optional)
   * @returns FieldDetection result
   */
  runFieldLevelDetection(
    field: Field,
    grayImage: cv.Mat,
    coloredImage?: cv.Mat
  ): FieldDetection {
    this.initializeFieldLevelAggregates(field);
    const fieldDetection = this.getFieldDetection(field, grayImage, coloredImage);
    this.updateAggregatesOnProcessedFieldDetection(field, fieldDetection);
    return fieldDetection;
  }

  /**
   * Update all aggregates when a field detection has been processed.
   *
   * @param field - Field that was processed
   * @param fieldDetection - Detection result
   */
  updateAggregatesOnProcessedFieldDetection(
    field: Field,
    fieldDetection: FieldDetection
  ): void {
    this.updateFieldLevelAggregatesOnProcessedFieldDetection(field, fieldDetection);

    const fieldLevelAggregates = this.getFieldLevelAggregates()!;

    this.updateFileLevelAggregatesOnProcessedFieldDetection(
      field,
      fieldDetection,
      fieldLevelAggregates
    );

    this.updateDirectoryLevelAggregatesOnProcessedFieldDetection(
      field,
      fieldDetection,
      fieldLevelAggregates
    );
  }

  /**
   * Update field-level aggregates when a field detection has been processed.
   *
   * @param field - Field that was processed
   * @param _fieldDetection - Detection result (not used in base implementation)
   */
  updateFieldLevelAggregatesOnProcessedFieldDetection(
    field: Field,
    _fieldDetection: FieldDetection
  ): void {
    this.updateFieldLevelAggregatesOnProcessedField(field);
  }

  /**
   * Update file-level aggregates when a field detection has been processed.
   *
   * @param field - Field that was processed
   * @param _fieldDetection - Detection result (not used in base implementation)
   * @param fieldLevelAggregates - Field-level aggregates
   */
  updateFileLevelAggregatesOnProcessedFieldDetection(
    field: Field,
    _fieldDetection: FieldDetection,
    fieldLevelAggregates: FieldLevelAggregates
  ): void {
    this.updateFileLevelAggregatesOnProcessedField(field, fieldLevelAggregates);
    // TODO: detection confidence metrics?
  }

  /**
   * Update directory-level aggregates when a field detection has been processed.
   *
   * @param field - Field that was processed
   * @param _fieldDetection - Detection result (not used in base implementation)
   * @param fieldLevelAggregates - Field-level aggregates
   */
  updateDirectoryLevelAggregatesOnProcessedFieldDetection(
    field: Field,
    _fieldDetection: FieldDetection,
    fieldLevelAggregates: FieldLevelAggregates
  ): void {
    this.updateDirectoryLevelAggregatesOnProcessedField(field, fieldLevelAggregates);
    // TODO: (if needed) update_directory_level_aggregates_on_processed_field
  }
}

/**
 * Template-level detection pass.
 *
 * Manages detection aggregates at the template level, coordinating
 * multiple field detection type passes.
 */
export class TemplateDetectionPass extends FilePassAggregates {
  /**
   * Initialize directory-level aggregates with field detection type tracking.
   *
   * @param initialDirectoryPath - Path to the initial directory
   * @param allFieldDetectionTypes - All field detection types in the template
   */
  initializeDirectoryLevelAggregates(
    initialDirectoryPath: string,
    allFieldDetectionTypes: string[]
  ): void {
    super.initializeDirectoryLevelAggregates(initialDirectoryPath);
    this.insertDirectoryLevelAggregates({
      files_by_label_count: new StatsByLabel('processed', 'multi_marked'),
      field_detection_type_wise_aggregates: Object.fromEntries(
        allFieldDetectionTypes.map((key) => [
          key,
          { fields_count: new StatsByLabel('processed') },
        ])
      ),
    });
  }

  /**
   * Initialize file-level aggregates with field detection type tracking.
   *
   * @param filePath - Path to the file being processed
   * @param allFieldDetectionTypes - All field detection types in the template
   */
  initializeFileLevelAggregates(
    filePath: string,
    allFieldDetectionTypes: string[]
  ): void {
    super.initializeFileLevelAggregates(filePath);
    this.insertFileLevelAggregates({
      field_detection_type_wise_aggregates: Object.fromEntries(
        allFieldDetectionTypes.map((key) => [
          key,
          { fields_count: new StatsByLabel('processed') },
        ])
      ),
    });
  }

  /**
   * Update all aggregates when a field detection has been processed.
   *
   * @param field - Field that was processed
   * @param fieldDetection - Detection result
   */
  updateAggregatesOnProcessedFieldDetection(
    field: Field,
    fieldDetection: FieldDetection
  ): void {
    this.updateFieldLevelAggregatesOnProcessedFieldDetection(field, fieldDetection);

    // TODO: can we move this getter to parent class?
    const fieldLevelAggregates = this.getFieldLevelAggregates()!;

    this.updateFileLevelAggregatesOnProcessedFieldDetection(
      field,
      fieldDetection,
      fieldLevelAggregates
    );
    this.updateDirectoryLevelAggregatesOnProcessedFieldDetection(
      field,
      fieldDetection,
      fieldLevelAggregates
    );
  }

  /**
   * Update field-level aggregates when a field detection has been processed.
   *
   * @param field - Field that was processed
   * @param _fieldDetection - Detection result (not used)
   */
  updateFieldLevelAggregatesOnProcessedFieldDetection(
    field: Field,
    _fieldDetection: FieldDetection
  ): void {
    this.updateFieldLevelAggregatesOnProcessedField(field);
  }

  /**
   * Update file-level aggregates when a field detection has been processed.
   *
   * @param field - Field that was processed
   * @param _fieldDetection - Detection result (not used)
   * @param fieldLevelAggregates - Field-level aggregates
   */
  updateFileLevelAggregatesOnProcessedFieldDetection(
    field: Field,
    _fieldDetection: FieldDetection,
    fieldLevelAggregates: FieldLevelAggregates
  ): void {
    this.updateFileLevelAggregatesOnProcessedField(field, fieldLevelAggregates);
    // TODO: detection confidence metrics?
  }

  /**
   * Update directory-level aggregates when a field detection has been processed.
   *
   * @param field - Field that was processed
   * @param _fieldDetection - Detection result (not used)
   * @param fieldLevelAggregates - Field-level aggregates
   */
  updateDirectoryLevelAggregatesOnProcessedFieldDetection(
    field: Field,
    _fieldDetection: FieldDetection,
    fieldLevelAggregates: FieldLevelAggregates
  ): void {
    this.updateDirectoryLevelAggregatesOnProcessedField(field, fieldLevelAggregates);
    const fieldDetectionType = field.fieldDetectionType;

    const directoryLevelAggregates = this.getDirectoryLevelAggregates()!;
    const fieldDetectionTypeWiseAggregates = (
      directoryLevelAggregates.field_detection_type_wise_aggregates as Record<
        string,
        { fields_count: StatsByLabel }
      >
    )[fieldDetectionType];
    // Update the processed field count for that runner
    fieldDetectionTypeWiseAggregates.fields_count.push('processed');
  }

  /**
   * Run field-level detection with automatic initialization of aggregates.
   *
   * Note: This method expects the field_detection to already be computed
   * by the field type runner. It initializes aggregates and updates them.
   *
   * @param field - Field that was processed
   * @param fieldDetection - Detection result (already computed)
   */
  runFieldLevelDetection(field: Field, fieldDetection: FieldDetection): void {
    this.initializeFieldLevelAggregates(field);
    this.updateAggregatesOnProcessedFieldDetection(field, fieldDetection);
  }

  /**
   * Update aggregates when a file has been processed.
   * Overrides base method to handle field detection type file runners.
   *
   * @param filePath - Path to the processed file
   * @param fieldDetectionTypeFileRunners - Map of field detection type to file runners
   */
  updateAggregatesOnProcessedFile(
    filePath: string,
    fieldDetectionTypeFileRunners?: Record<string, { getFileLevelDetectionAggregates(): unknown }>
  ): void {
    super.updateAggregatesOnProcessedFile(filePath);

    if (!fieldDetectionTypeFileRunners) {
      return;
    }

    const fileLevelAggregates = this.getFileLevelAggregates()!;
    const fieldDetectionTypeWiseAggregates = fileLevelAggregates.field_detection_type_wise_aggregates as Record<
      string,
      unknown
    >;

    for (const fieldDetectionTypeFileRunner of Object.values(fieldDetectionTypeFileRunners)) {
      const fieldDetectionType = (fieldDetectionTypeFileRunner as {
        fieldDetectionType: string;
        getFileLevelDetectionAggregates(): unknown;
      }).fieldDetectionType;
      fieldDetectionTypeWiseAggregates[fieldDetectionType] =
        fieldDetectionTypeFileRunner.getFileLevelDetectionAggregates();
    }

    // Populate typed field results from repository
    // Get repository from any field type runner (they all share the same repository)
    const anyRunner = Object.values(fieldDetectionTypeFileRunners)[0] as {
      repository?: DetectionRepository;
    };
    if (anyRunner?.repository) {
      try {
        const fileResults = anyRunner.repository.getFileResults(filePath);

        // Map all field types by field_label for interpretation access
        const bubbleFieldsByLabel: Record<string, BubbleFieldDetectionResult> = {};
        for (const result of fileResults.bubbleFields.values()) {
          bubbleFieldsByLabel[result.fieldLabel] = result;
        }

        const ocrFieldsByLabel: Record<string, OCRFieldDetectionResult> = {};
        for (const result of fileResults.ocrFields.values()) {
          ocrFieldsByLabel[result.fieldLabel] = result;
        }

        const barcodeFieldsByLabel: Record<string, BarcodeFieldDetectionResult> = {};
        for (const result of fileResults.barcodeFields.values()) {
          barcodeFieldsByLabel[result.fieldLabel] = result;
        }

        const fileLevelAggregates = this.getFileLevelAggregates()!;
        fileLevelAggregates.bubble_fields = bubbleFieldsByLabel;
        fileLevelAggregates.ocr_fields = ocrFieldsByLabel;
        fileLevelAggregates.barcode_fields = barcodeFieldsByLabel;
      } catch (error) {
        // Check if it's a KeyError-equivalent (file not found in repository)
        if (error instanceof Error && error.message.includes('No results found for file')) {
          // File not yet finalized in repository - this should not happen
          // but if it does, log and re-raise to surface the issue
          const logger = new Logger('TemplateDetectionPass');
          logger.error(
            `File ${filePath} not found in repository after finalize_file(). ` +
              `This indicates a bug in the repository lifecycle management. Error: ${error.message}`
          );
          // Re-raise to surface the issue during development
          throw error;
        } else {
          // Other unexpected errors
          const logger = new Logger('TemplateDetectionPass');
          logger.error(
            `Unexpected error populating field results from repository for ${filePath}: ${error}`
          );
          // Initialize empty dicts to prevent KeyError in interpretation
          const fileLevelAggregates = this.getFileLevelAggregates()!;
          if (!fileLevelAggregates.bubble_fields) {
            fileLevelAggregates.bubble_fields = {};
          }
          if (!fileLevelAggregates.ocr_fields) {
            fileLevelAggregates.ocr_fields = {};
          }
          if (!fileLevelAggregates.barcode_fields) {
            fileLevelAggregates.barcode_fields = {};
          }
        }
      }
    }
  }
}

